packages/aws-cdk/lib/api/hotswap-deployments.ts

```diff
@@ -0,0 +1,100 @@
+import * as cfn_diff from '@aws-cdk/cloudformation-diff';
+import * as cxapi from '@aws-cdk/cx-api';
+import { CloudFormation } from 'aws-sdk';
+import { ISDK, Mode, SdkProvider } from './aws-auth';
+import { DeployStackResult } from './deploy-stack';
+import { HotswapOperation, ListStackResources } from './hotswap/common';
+import { isHotswappableLambdaFunctionChange } from './hotswap/lambda-functions';
+import { CloudFormationStack } from './util/cloudformation';
+
+/**
+ * Perform a hotswap deployment,
+ * short-circuiting CloudFormation if possible.
+ * If it's not possible to short-circuit the deployment
+ * (because the CDK Stack contains changes that cannot be deployed without CloudFormation),
+ * returns `undefined`.
+ */
+export async function tryHotswapDeployment(
+  sdkProvider: SdkProvider, assetParams: { [key: string]: string },
+  cloudFormationStack: CloudFormationStack, stackArtifact: cxapi.CloudFormationStackArtifact,
+): Promise<DeployStackResult | undefined> {
+  const currentTemplate = await cloudFormationStack.template();
+  const stackChanges = cfn_diff.diffTemplate(currentTemplate, stackArtifact.template);
+
+  // resolve the environment, so we can substitute things like AWS::Region in CFN expressions
+  const resolvedEnv = await sdkProvider.resolveEnvironment(stackArtifact.environment);
+  const hotswappableChanges = findAllHotswappableChanges(stackChanges, {
+    ...assetParams,
+    'AWS::Region': resolvedEnv.region,
+    'AWS::AccountId': resolvedEnv.account,
+  });
+  if (!hotswappableChanges) {
+    // this means there were changes to the template that cannot be short-circuited
+    return undefined;
+  }
+
+  // create a new SDK using the CLI credentials, because the default one will not work for new-style synthesis -
+  // it assumes the bootstrap deploy Role, which doesn't have permissions to update Lambda functions
+  const sdk = await sdkProvider.forEnvironment(resolvedEnv, Mode.ForWriting);
+  // apply the short-circuitable changes
+  await applyAllHotswappableChanges(sdk, stackArtifact, hotswappableChanges);
+
+  return { noOp: hotswappableChanges.length === 0, stackArn: cloudFormationStack.stackId, outputs: cloudFormationStack.outputs, stackArtifact };
+}
+
+function findAllHotswappableChanges(
```

Comment:
Do you mind if we skip this for now? In my opinion, this is introducing a lot of complexity that we don't need right now. For example, determining what to do with a given change after getting an array of sub-results for it is actually non-trivial (if all of them are strings, then if there's at least one `'impossible'`, then the combined result for the change is `'impossible`, otherwise `'irrelevant'`; if there's one `HotswapOperation`, then the combined result is that `HotswapOperation`; if there's more than one `HotswapOperation`... who know? Error out? Apply all of them?).

We'll make sure to factor out the part that deals with `CDKMetadata` from the detectors when we extract the plugin architecture for adding more resource types.

Reply:
I'm fine with leaving this as is if we are likely to redesign it in the near future for the plugin capabilities